Add support for property based assertion

Very often it's required to assert multiple properties of a given object and if the assertion fails, you have no clue on which property the assertion failed, unless of course you use the nice "Clues" feature of Kotest :smile:. Writing clues for each property however feels a bit verbose and clunky and it's not type safe if the property is renamed for example.

Here an example to show the problem:
```kotlin
data class Person(val firstName: String, val lastName: String, val age: Int)

val expected = Person("James", "Bond", 56)
val actual = Person("James", "Bond", 52)

withClue("Unexpected firstName") { actual.firstName shouldBe expected.firstName }
withClue("Unexpected lastName") { actual.lastName shouldBe expected.lastName }
withClue("Unexpected age") { actual.age shouldBe expected.age }
```

Of course, one could simply assert the whole object (especially if it's a data class) like so:
```kotlin
actual shouldBe expected
```

But what if I have many properties in my class and I don't want to assert all properties because not all are relevant for my test case. Plus, what if I have many properties and only one fails? I would get a huge string generated for each object and it wouldn't be that easy to find which property is different..

So, I tried to find something in the API that could help me with this problem but couldn't find anything, so I ended up creating the following extension function:
```kotlin
/**
 * Extends Kotest's [shouldBe] to work with properties of type [KProperty0],
 * i.e. property references on instances, so that the name of the property
 * can be printed as clue if the assertion of the property's values fail.
 *
 * Example:
 * ```
 * data class MyValue(val prop: Int)
 *
 * val expected = MyValue(1)
 * val passes = MyValue(1)
 * val fails = MyValue(0)
 *
 * // will pass
 * passes::prop shouldGet expected::prop
 * // will fail with message:
 * // Unexpected prop
 * // expected:<1> but was:<0>
 * fails::prop shouldGet expected::prop
 * ```
 */
infix fun <V> KProperty0<V>.shouldGet(expected: KProperty0<V>) {
    withClue("Unexpected $name") {
        get() shouldBe expected.get()
    }
}
```

As you may have guessed from the code and KDoc comment, this allows me to write the code above as follows and yield the exact same result:
```kotlin
data class Person(val firstName: String, val lastName: String, val age: Int)

val expected = Person("James", "Bond", 56)
val actual = Person("James", "Bond", 52)

actual::firstName shouldGet expected::firstName
actual::lastName shouldGet expected::lastName
actual::age shouldGet expected::age
```

To make it really nice for bigger classes, one could do something like that:
```kotlin
withClue("Assertion for Person failed") {
    assertSoftly {
        actual::firstName shouldGet expected::firstName
        actual::lastName shouldGet expected::lastName
        actual::age shouldGet expected::age
    }
}
```

Which brings me to further ideas like having an API where I can pass the actual and expected objects and a list of properties and it generates the clue with the class name like above.

Anyway, this is not necessarily a feature request but just wanted to share this use-case with this forum to check if there is a bigger demand for something like this and let you guys consider adding such concepts to the API in future releases. I'm happy to help with the coding part if required.

OK, I think I've just found a a bug in `shouldHaveValue`. It doesn't print the property information if it's used inside an `assertSoftly` block. On the other hand, it seems to print the expected and actual values twice if not within an `assertSoftly` block.

This is how I could reproduce it:

**Wihtout** `assertSoftly` block:
```kotlin
data class Person(val firstName: String, val lastName: String, val age: Int)

val expected = Person("James", "Bond", 56)
val actual = Person("John", "Bond", 52)

actual::firstName shouldHaveValue expected.firstName
actual::lastName shouldHaveValue expected.lastName
actual::age shouldHaveValue expected.age
```
Prints:
```plain
io.kotest.assertions.AssertionFailedError: Property 'firstName' should have value James
expected:<"James"> but was:<"John">
expected:<"James"> but was:<"John">
```


**With** `assertSoftly` block:
```kotlin
data class Person(val firstName: String, val lastName: String, val age: Int)

val expected = Person("James", "Bond", 56)
val actual = Person("John", "Bond", 52)

withClue("Assertion for Person failed") {
    assertSoftly {
        actual::firstName shouldHaveValue expected.firstName
        actual::lastName shouldHaveValue expected.lastName
        actual::age shouldHaveValue expected.age
    }
}
```
Prints:
```plain
io.kotest.assertions.MultiAssertionError: The following 2 assertions failed:
1) Assertion for Person failed
expected:<"James"> but was:<"John">
   at PropertiesKt$haveValue$1.test(properties.kt:19)
2) Assertion for Person failed
expected:<56> but was:<52>
   at PropertiesKt$haveValue$1.test(properties.kt:19)
```

As you can see the version with the soft block is missing the message `Property 'firstName' should have value James`.

I had a look at the code and the problem seems to be in the `runCatching{ }` block in the `haveValue()` function (Version 5.8.0, file PropertiesKt, line 18). When running in soft mode, no exception is thrown, so the part further down that composes the message with the property details is never reached. I suspect the cause of the duplicated message when not in soft mode has also to do with it but didn't investigate further.

--- a/documentation/docs/assertions/core.md
+++ b/documentation/docs/assertions/core.md
@@ -13,6 +13,7 @@ Matchers provided by the `kotest-assertions-core` module.
 | General                                 |                                                                                                  |
 |-----------------------------------------|--------------------------------------------------------------------------------------------------|
 | `obj.shouldBe(other)`                   | General purpose assertion that the given obj and other are both equal                            |
+| `obj::prop.shouldHaveValue(other)`      | General purpose assertion on a property value printing information of the property on failure.   |
 | `expr.shouldBeTrue()`                   | Convenience assertion that the expression is true. Equivalent to `expr.shouldBe(true)`           |
 | `expr.shouldBeFalse()`                  | Convenience assertion that the expression is false. Equivalent to `expr.shouldBe(false)`         |
 | `shouldThrow<T> { block }`              | General purpose construct that asserts that the block throws a `T` Throwable or a subtype of `T` |

--- a/kotest-assertions/kotest-assertions-core/api/kotest-assertions-core.api
+++ b/kotest-assertions/kotest-assertions-core/api/kotest-assertions-core.api
@@ -1,9 +1,3 @@
-public final class PropertiesKt {
-	public static final fun haveValue (Ljava/lang/Object;)Lio/kotest/matchers/Matcher;
-	public static final fun shouldHaveValue (Lkotlin/reflect/KProperty0;Ljava/lang/Object;)V
-	public static final fun shouldMatch (Lkotlin/reflect/KProperty0;Lkotlin/jvm/functions/Function1;)V
-}
-
 public final class io/kotest/assertions/nondeterministic/ContinuallyConfiguration {
 	public synthetic fun <init> (JJLio/kotest/assertions/nondeterministic/DurationFn;Lkotlin/jvm/functions/Function3;Lkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public final fun component1-UwyO8pc ()J
@@ -1834,6 +1828,13 @@ public final class io/kotest/matchers/paths/PathsKt {
 	public static final fun startWithPath (Ljava/io/File;)Lio/kotest/matchers/Matcher;
 	public static final fun startWithPath (Ljava/lang/String;)Lio/kotest/matchers/Matcher;
 	public static final fun startWithPath (Ljava/nio/file/Path;)Lio/kotest/matchers/Matcher;
+}
+
+public final class io/kotest/matchers/properties/PropertiesKt {
+	public static final fun haveValue (Ljava/lang/Object;)Lio/kotest/matchers/Matcher;
+	public static final fun shouldHaveValue (Lkotlin/reflect/KProperty0;Ljava/lang/Object;)V
+	public static final fun shouldMatch (Lkotlin/reflect/KProperty0;Lkotlin/jvm/functions/Function1;)V
+	public static final fun shouldNotHaveValue (Lkotlin/reflect/KProperty0;Ljava/lang/Object;)V
 }
 
 public final class io/kotest/matchers/property/MatchersKt {

--- a/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/properties/properties.kt
+++ b/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/properties/properties.kt
@@ -1,37 +1,45 @@
+package io.kotest.matchers.properties
+
+import io.kotest.assertions.Actual
+import io.kotest.assertions.Expected
+import io.kotest.assertions.eq.eq
+import io.kotest.assertions.intellijFormatError
+import io.kotest.assertions.print.print
 import io.kotest.assertions.withClue
-import io.kotest.matchers.EqualityMatcherResult
 import io.kotest.matchers.Matcher
 import io.kotest.matchers.MatcherResult
 import io.kotest.matchers.should
 import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNot
+import io.kotest.matchers.shouldNotBe
 import kotlin.reflect.KProperty0
 
 /**
- * Assert that this property has a specific value. Unlike regular [shouldBe], name of the property will be
- * automatically added to the error message
+ * Asserts that this property has a specific value. Unlike regular [shouldBe], the name of the property
+ * will be automatically added to the error message
  */
-infix fun <T> KProperty0<T>.shouldHaveValue(t: T) = this should haveValue(t)
+infix fun <T> KProperty0<T>.shouldHaveValue(expected: T) = this should haveValue(expected)
 
-fun <T> haveValue(t: T) = object : Matcher<KProperty0<T>> {
+/**
+ * Asserts that this property does not have a specific value. Unlike regular [shouldNotBe], the name of the
+ * property will be automatically added to the error message
+ */
+infix fun <T> KProperty0<T>.shouldNotHaveValue(expected: T) = this shouldNot haveValue(expected)
+
+fun <T> haveValue(expected: T) = object : Matcher<KProperty0<T>> {
    override fun test(value: KProperty0<T>): MatcherResult {
+      val prependMessage = { "Assertion failed for property '${value.name}'" }
       val actual = value.get()
-      val res = runCatching {
-         actual shouldBe t
-      }
+      return object : MatcherResult {
+         override fun passed(): Boolean =
+            eq(actual, expected) == null
 
-      return EqualityMatcherResult(
-         res.isSuccess,
-         actual,
-         t,
-         {
-            val detailedMessage = res.exceptionOrNull()?.message
-            "Property '${value.name}' should have value $t\n$detailedMessage"
-         },
-         {
-            val detailedMessage = res.exceptionOrNull()?.message
-            "Property '${value.name}' should not have value $t\n$detailedMessage"
-         },
-      )
+         override fun failureMessage(): String =
+            prependMessage() + "\n" + intellijFormatError(Expected(expected.print()), Actual(actual.print()))
+
+         override fun negatedFailureMessage(): String =
+            prependMessage() + "\n${expected.print().value} should not equal ${actual.print().value}"
+      }
    }
 }
 

--- a/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/io/kotest/matchers/properties/PropertiesKtTest.kt
+++ b/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/io/kotest/matchers/properties/PropertiesKtTest.kt
@@ -1,11 +1,12 @@
 package io.kotest.matchers.properties
 
+import io.kotest.assertions.assertSoftly
 import io.kotest.assertions.throwables.shouldThrowAny
+import io.kotest.assertions.withClue
 import io.kotest.core.spec.style.FunSpec
 import io.kotest.matchers.shouldBe
 import io.kotest.matchers.string.shouldContain
 import io.kotest.matchers.string.shouldStartWith
-import shouldHaveValue
 
 class PropertiesKtTest : FunSpec({
 
@@ -16,14 +17,100 @@ class PropertiesKtTest : FunSpec({
    test("KProperty0<T>.shouldHaveValue with error message") {
       shouldThrowAny {
          Foo("1")::a shouldHaveValue "2"
-      }.message shouldStartWith "Property 'a' should have value 2"
+      }.message shouldBe
+         """
+         Assertion failed for property 'a'
+         expected:<"2"> but was:<"1">
+         """.trimIndent()
    }
 
-   test("KProperty0<T>.shouldHaveValue with error message should include see-difference formatting") {
-      shouldThrowAny {
-         Foo("1")::a shouldHaveValue "2"
-      }.message shouldContain "expected:<\"2\"> but was:<\"1\">"
+   test("KProperty0<T>.shouldNotHaveValue happy path") {
+      Foo("1")::a shouldNotHaveValue "2"
    }
+
+   test("KProperty0<T>.shouldNotHaveValue with error message") {
+      shouldThrowAny {
+         Foo("1")::a shouldNotHaveValue "1"
+      }.message shouldBe
+         """
+         Assertion failed for property 'a'
+         "1" should not equal "1"
+         """.trimIndent()
+   }
+
+   test("KProperty0<T>.shouldHaveValue with clue in the error message") {
+      shouldThrowAny {
+         withClue("This is a clue") {
+            Foo("1")::a shouldHaveValue "2"
+         }
+      }.message shouldBe
+         """
+         This is a clue
+         Assertion failed for property 'a'
+         expected:<"2"> but was:<"1">
+         """.trimIndent()
+   }
+
+   test("KProperty0<T>.shouldHaveValue with assertSoftly happy path") {
+      val value = Bar("1", 2, 3)
+      assertSoftly {
+         value::foo shouldHaveValue "1"
+         value::bar shouldHaveValue 2
+         value::bla shouldHaveValue 3
+      }
+   }
+
+   test("KProperty0<T>.shouldHaveValue with assertSoftly and error message") {
+      val value = Bar("1", 2, 3)
+      shouldThrowAny {
+         assertSoftly {
+            value::foo shouldHaveValue "1"
+            value::bar shouldHaveValue 1
+            value::bla shouldHaveValue 2
+         }
+      }.message.also {
+         it shouldStartWith "The following 2 assertions failed:"
+         it shouldContain
+            """
+            1) Assertion failed for property 'bar'
+            expected:<1> but was:<2>
+            """.trimIndent()
+         it shouldContain
+            """
+            2) Assertion failed for property 'bla'
+            expected:<2> but was:<3>
+            """.trimIndent()
+      }
+   }
+
+
+   test("KProperty0<T>.shouldHaveValue with assertSoftly, clue and error message") {
+      val value = Bar("1", 2, 3)
+      shouldThrowAny {
+         withClue("This is a clue") {
+            assertSoftly {
+               value::foo shouldHaveValue "1"
+               value::bar shouldHaveValue 1
+               value::bla shouldHaveValue 2
+            }
+         }
+      }.message.also {
+         it shouldStartWith "The following 2 assertions failed:"
+         it shouldContain """
+1) This is a clue
+Assertion failed for property 'bar'
+expected:<1> but was:<2>
+         """.trimIndent()
+         it shouldContain """
+2) This is a clue
+Assertion failed for property 'bla'
+expected:<2> but was:<3>
+         """.trimIndent()
+      }
+   }
+
 })
 
 data class Foo(val a: String)
+
+data class Bar(val foo: String, val bar: Int, val bla: Int)

